### DISCUSSION TITLE
Enhance the Trivy plugin index

### DISCUSSION THREAD

**Discussion by knqyf263** [aquasecurity/trivy-plugin-index#2] [Development]
## Overview
The current schema for the trivy-plugin-index is very simple:

```yaml
name: count
repository: github.com/aquasecurity/trivy-plugin-count
```

This simplicity is due to the presence of `plugin.yaml` in the plugin repositories, which is automatically crawled by CI to extract necessary information and deploy it using GitHub Pages: [index.yaml](https://aquasecurity.github.io/trivy-plugin-index/v1/index.yaml).

### Versioning
The YAML itself does not include versioning; instead, the HTTP path has a `v1` prefix (https://aquasecurity.github.io/trivy-plugin-index/v1/index.yaml) to prevent Trivy from breaking due to destructive changes in the index schema. Revising the `index.yaml` schema might be beneficial, considering the potential inclusion of metadata in the future.

### Automatic Crawling
The advantage of automatic crawling is the elimination of duplicate definitions. However, there is a risk of breakage if incorrect changes are made to `plugin.yaml` in the plugin repository. On the other hand, if `plugin.yaml` breaks, it would also prevent installation via Trivy, so excluding it from the index might be beneficial.

### Tracking `index.yaml`
Currently, `index.yaml` is [automatically generated on CI](https://github.com/aquasecurity/trivy-plugin-index/blob/a91783ab7be56982ded575848113766779ffc01a/.github/workflows/deploy-index.yaml#L24-L25) and deployed to GitHub Pages. There should be a discussion on whether this file should be managed with Git.

### Security
Security is a crucial consideration. The plugin repository may change, posing risks. However, this risk is inherent to the plugins themselves, as stated in the [documentation](https://aquasecurity.github.io/trivy/v0.51/docs/advanced/plugins/), and not due to the automatic crawling of the index.

**Comment by knqyf263** [comment]
@DmitriyLewen What do you think?

**Reply by DmitriyLewen** [reply]
> However, there is a risk of breakage if incorrect changes are made to plugin.yaml in the plugin repository.

I think we need to leave automatic crawling. But you are right - it can be problem.

To eliminate broken plugins, I propose the following idea - update `test` workflow:
- check required fields in `plugin.yaml` (for `trivy-plugin-index`).
- try to download plugin and check its fields
- checks the availability of the `uri` and `bin` fields.

> Tracking index.yaml

I would add `index.yaml` to this repo.
Our team will need to support `trivy-plugin-index`. Therefore, it will be more convenient if it is stored in repository (e.g. you can view change history).
It’s also easier for me to look at `yaml` file than to open website and look for a repository in it. I think I'm not the only one 😃 

> The plugin repository may change

We can check the availability of the repository when we update the indexes. 
UPD: or we can add crow workflow to check availability of plugin repositories.

**Reply by knqyf263** [reply]
I added a workflow for automatically creating a PR to update `index.yaml`. Once the PR is merged, the index will be deployed. @DmitriyLewen Please take a look.
https://github.com/aquasecurity/trivy-plugin-index/pull/6
https://github.com/aquasecurity/trivy-plugin-index/actions/runs/9076122027

**Reply by knqyf263** [reply]
Also, you can see the history now.
https://github.com/aquasecurity/trivy-plugin-index/blob/main/index.yaml

**Reply by DmitriyLewen** [reply]
We will need to merg 2 PRs - add plugin to `plugins` directory + update `index.json`.
Do you have reason why you don't want to just push commit with updated `index.json` after merging PR with new/removed/updated plugin?

**Reply by knqyf263** [reply]
I thought we wanted to review the generated index manually. Do you want to automate it?

**Reply by DmitriyLewen** [reply]
> Do you want to automate it?

We will need to check new/updated/removed plugin (2 lines (name + repository)). So I think, yes, we can automate this.

**Reply by knqyf263** [reply]
How about this workflow?
https://github.com/aquasecurity/trivy-plugin-index/commit/cf3c352c84c3bdbd523da9e268432fdc5e272a75

**Reply by DmitriyLewen** [reply]
looks good 👍

**Comment by knqyf263** [comment]
I learned from that mistake and decided to add metadata this time, as Trivy initially had a schema that could not have metadata, which required destructive changes to the schema later on.
https://github.com/aquasecurity/trivy-plugin-index/commit/7f5002adcd62dbc87b19b7e046ef051b73684b64

**Reply by DmitriyLewen** [reply]
Agree with your idea.

> https://aquasecurity.github.io/trivy-plugin-index/v1/index.yaml

Perhaps we can rename `v1` to `index` (as example) to avoid confusion if we will use version `2` but store file in `v1` directory.

**Reply by knqyf263** [reply]
Regardless of the file schema, we should have a prefix in the endpoint. Otherwise, an older Trivy will be broken with v2.

**Reply by knqyf263** [reply]
With version prefix:

- New Trivy with v2 support: Access `https://aquasecurity.github.io/trivy-plugin-index/v2/index.yaml`
- Old Trivy with only v1 support: Access `https://aquasecurity.github.io/trivy-plugin-index/v1/index.yaml`.

Without version prefix:

- New Trivy with v2 support: Access `https://aquasecurity.github.io/trivy-plugin-index/index/index.yaml` successfully
- Old Trivy with only v1 support: Access `https://aquasecurity.github.io/trivy-plugin-index/index/index.yaml` and fail

Another benefit of having a version in the index YAML is that people can probably use any HTTP path when they use [custom indexes](https://krew.sigs.k8s.io/docs/developer-guide/custom-indexes/) (Trivy doesn't yet support it, though).

**Reply by DmitriyLewen** [reply]
Got it. You are right - we need to use version prefix 👍
